osquery/tables/applications/posix/prometheus_metrics.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <sstream>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <osquery/tables/applications/posix/prometheus_metrics.h>
+#include <osquery/tables/applications/posix/prometheus_metrics_utils.h>
+
+namespace http = boost::network::http;
+
+namespace osquery {
+namespace tables {
+
+void PrometheusMetrics::parseScrapeResults() {
+  for (auto const& target : scrapeResults_) {
+    std::stringstream ss(target.second->content);
+    std::string dest;
+    std::string ts(std::to_string(target.second->timestampMS.count()));
+
+    while (std::getline(ss, dest)) {
+      if (dest[0] != '#') {
+        std::stringstream iss(dest);
+        std::string idest;
+        std::vector<std::string> metric;
+
+        while (std::getline(iss, idest, ' ')) {
+          if (dest != "") {
+            metric.push_back(idest);
+          }
+        }
+
+        if (metric.size() > 1) {
+          Row r;
+          r[col_target_name] = target.first;
+          r[col_timestamp] = ts;
+          r[col_metric] = metric[0];
+          r[col_value] = metric[1];
+
+          rows_.push_back(r);
+        }
+      }
+    }
+  }
+}
+
+void PrometheusMetrics::scrapeTargets() {
+  for (auto& target : scrapeResults_) {
+    try {
+      http::client::request request(target.first);
+      http::client::response response(client_.get(request));
+
+      target.second->timestampMS =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::system_clock::now().time_since_epoch());
+      target.second->content = static_cast<std::string>(body(response));
+
+    } catch (std::exception& e) {
+      LOG(ERROR) << "failed on scrape of target " << target.first
+                 << " with error: " << e.what();
+    }
+  }
+}
+
+QueryData& PrometheusMetrics::queryPrometheusTargets() {
+  scrapeTargets();
+  parseScrapeResults();
```

Comment:
This function that relies entirely on side-effects is hard to follow. Instead, try to structure it so that you pass things to a function and use what the function returns. This makes code easier to follow and debug. Could you pass scrapeResults to scrapeTargets as a parameter? Same with parseScrapeResults?